doc/developer/design/20210311_prometheus_source.md

```diff
@@ -0,0 +1,141 @@
+# Summary
+
+This design doc introduces a way to import data from our internal prometheus registry into materialized views that can be inspected with SQL commands.
+
+# Goals
+
+* Allow our users to read & make decisions based on our metrics data without setting up a whole metrics ingestion/presentation pipeline.
+
+* Let administrators inspect that metrics data (as much as metrics' attached data allows) together with the other available data we present in system tables.
+
+* Retain a long-enough history of metrics values to allow making decisions based on how the metrics values change (but short enough to not consume unbounded memory).
+
+# Non-Goals
+
+None of the following are intended to be done with this design:
+
+* Export our system table logs as prometheus (yet) -- this design covers only the "metrics import" direction.
+
+* Expose any `CREATE SOURCE` mechanics for importing from arbitrary/external prometheus endpoints.
+
+* Come up with a data retention scheme other than "older than n minutes gets dropped". If this was a full-featured time series database design, we'd want to thin metrics into minute/hour/day buckets. This isn't that.
+
+# Description
+
+<!--
+    Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+    If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+By implementing this design, we'll make the following changes to materialized:
+
+1. Introduce a background task that regularly scrapes the internal metrics registry using the registry's [`gather`](https://docs.rs/prometheus/0.12.0/prometheus/struct.Registry.html#method.gather) method (currently we can use the global registry, but later we'll need a handle to the ["materialized" metrics registry][#5825]; the mechanism will remain the same).
+2. Ingest the generated metrics into a dataflow that populates metrics into tables like the following:
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+    value float8
+);
+
+CREATE TABLE metrics_table_meta (
+    metric TEXT,
+    metric_type TEXT,
+    help TEXT
+);
+```
+
+3. This not-quite-source polls the endpoint once a minute (via the configured the `poll_interval` duration), and retains metric readings for a number of minutes (strawman proposal: 5 minutes) to make it possible for users to visualize changes in the metrics as they would with a "full" metrics pipeline.
```

Comment:
I think that stating that we will expose this configuration parameter to end users should be left for some follow up work.

That said, if you're going to define it here, what exactly is the API? Is it a CLI arg or a SQL function or a system variable that can be SET? I think a CLI arg is probably the most straightforward thing to implement, and matches things like rdkafka log frequency.